python/ray/tune/trainable.py

```diff
@@ -394,6 +394,90 @@ def restore_from_object(self, obj):
         self.restore(checkpoint_path)
         shutil.rmtree(tmpdir)
 
+    @staticmethod
+    def get_checkpoint_path(checkpoint_dir):
+        """Given a directory of a specific checkpoint, returns the full
+        checkpoint path.
+
+        Args:
+            checkpoint_dir (str): Path pointing to a checkpoint directory.
+        Returns:
+            A string with the full path to the checkpoint within the directory.
+        """
+        if 'checkpoint' not in checkpoint_dir:
+            raise ValueError("The given path does not point to a specific "
+                             f"checkpoint: {checkpoint_dir}")
+        # if it directly points to the checkpoint file, just return it
+        if os.path.isfile(checkpoint_dir):
+            return checkpoint_dir
+
+        # else derive the checkpoint file and return it
+        checkpoint_number = checkpoint_dir.split('_')[-1]
+        return os.path.join(checkpoint_dir, f'checkpoint-{checkpoint_number}')
+
+    @staticmethod
+    def get_last_checkpoint(dir_path):
+        """Finds and returns last checkpoint from given directory.
+
+        Args:
+            dir_path (str): Path pointing to a directory with checkpoints.
+
+        Returns:
+            Last checkpoint in given directory.
+        """
+        # if dir_path already points to a specific checkpoint load it
+        if 'checkpoint' in dir_path:
+            return Trainable.get_checkpoint_path(dir_path)
+
+        # else load the last checkpoint from all available checkpoints
+        abs_dir_path = os.path.abspath(dir_path)
+        checkpoints = [
+            f for f in os.listdir(abs_dir_path) if f.startswith('checkpoint')
+        ]
+        if len(checkpoints) == 0:
+            raise ValueError(f"No checkpoints in directory: {dir_path}")
+
+        # sort according to checkpoint number after '_'
+        sorted_checkpoints = sorted(
+            checkpoints, key=lambda cp: int(cp.split('_')[-1]))
+        last_checkpoint_dir = os.path.join(abs_dir_path,
+                                           sorted_checkpoints[-1])
+        # retrieve checkpoint number for filename and construct full path
+        last_checkpoint_no = last_checkpoint_dir.split('_')[-1]
+        last_checkpoint_path = os.path.join(
+            last_checkpoint_dir, f'checkpoint-{last_checkpoint_no}')
```

Comment:
While passing the current training iteration as an index to `TrainableUtil.make_checkpoint_dir` it is not enforced and users can pass arbitrary strings here. Can we instead do something similar to the `get_best_checkpoint`, but just choose the checkpoint with the maximum number of iterations (Key stored in `ray.tune.result.TRAINING_ITERATION`)?

Reply:
Sounds good, I'll give it a try! 

How should I identify checkpoint directories and files in general? Can I rely on them starting with `checkpoint` or is there a more generic approach? Is the `checkpoint` prefix stored in some constant somewhere that I could access?

Reply:
The `checkpoint` prefix is currently hardcoded in `TrainableUtil.make_checkpoint_dir()` - since we're moving these functions into the same class you can just hardcode it and don't have to read it from a constant. 

I think it's fine to rely on the prefix, but be sure to call `find_checkpoint_dir()` to ensure it is actually a checkpoint. We could potentially get rid of the filter for `checkpoint_`, but since that prefix is currently actually enforced we don't need to do that right now.